### PR TITLE
Refactor batch classes to reduce abstract class complexity

### PR DESCRIPTION
**Pull request by zachwills**
Related to #61 

This is a WIP 'proposal' type of PR. I definitely went out on a bit of a limb and refactored the way that child classes of `Batch` are implemented.

My thought process while working on this was that I wanted to make the extension of the `Batch` class as straightforward as possible. Previously, we just relied on the singular `get_results()` function to be implemented on each child class. I think that approach isn't ideal because it puts a lot of responsibility on that singular function (it was setting args in addition to running the query).

I think the current approach outlined in this PR has a better separation of concerns which removes some of the complexity from the base abstract class. Rather than relying on multiple `switch` statements that would grow as we added the additional batch process types ( #42 ), we instead force the child classes to define these functions themselves.

I'm excited to get some feedback (even if it ends up being that this approach isn't ideal) so please don't hesitate to share your thoughts :)

PS: Was able to refactor this relatively easily thanks to all of our unit tests! Hooray for those 🎉

### CODE REVIEW

**Comment by zachwills** [comment]
Shoot I can't view the failed Scrutinizer test because I don't have access to it anymore. `phpunit` passes successfully locally so not sure where the hiccup is:

screen shot 2016-12-07 at 5 15 09 pm

**Comment by jjeaton** [comment]
@zachwills missing doc comments on class-batch-users.php.

We will probably migrate this to Travis soon for running tests, so Scrutinizer access won't be an issue going forward.

**Comment by zachwills** [comment]
@jjeaton ah ok. I left those comments off because they would basically have been copy / pastes of the function comment that's on the `abstract function xxx()` version of the function on the parent class. I'm guessing Scrutinizer is configured to fail the build if something doesn't match the WP coding standards?

I'll add in the comments to get the build passing. I think there _might_ be a case for excluding the required comments rule, because in instance like this it feels a bit counterproductive to implement near identical comments across the parent and child classes.

**Comment by JasonHoffmann** [comment]
Still learning my way around the codebase, but I do like this approach. Kind of defines the flow more in the base class, and then only the individual processing is left to individual classes.

I know there was some thoughts about normalizing variable / argument names a bit. Right now, the arguments are being processed in get_results. Would we want to pull that out into an extendable function as well?

**Comment by zachwills** [comment]
@JasonHoffmann thanks for the initial feedback :)

> I know there was some thoughts about normalizing variable / argument names a bit.

Yes- that's what #61 is for... but the more I think about it, I'm not sure if it's worth the extra logic in the code to normalize a few of the query arguments. Maybe we should list out the args we can think of that are different / should be normalized and then evaluate if it is worth it to normalize them? I'll ask on that issue's thread since the discussion should probably be handled there.

File: includes/abstracts/abstract-batch.php [hunk 1]
```diff
@@ -73,13 +88,60 @@
 	public $total_num_results;
 
 	/**
+	 * Get results function for the registered batch process.
+	 *
+	 * @return array
+	 */
+	abstract public function individual_get_results();
```
Comment (jjeaton) [inline includes/abstracts/abstract-batch.php]: what does the `individual_` naming convention represent? An individual batch? Maybe `batch_` would be a better prefix in that case? 🌴

**Comment by zachwills** [comment]
@jjeaton I updated the function names- let me know what you think.

**Comment by jjeaton** [comment]
LGTM!

**Pull request merged**